[Not For Landing] Main Wholesale reformatting discussion

pmatos opened the pull request:
This is related to #3412 

It contains a full tree format, this is formatting:
* FEXCore
* FEXHeaderUtils
* include
* Scripts
* Source
* ThunkLibs

It's formatting using a clang-format calculated through clang-unformat from the FEXCore directory - i.e. specifically trying to minimize changes there. 

Let me know what you think. We can use this PR to get suggestions on formatting improvements, I can then tweak the clang-format and push new versions of the tree until we are happy.

neobrain commented on FEXCore/Source/Interface/HLE/Thunks/Thunks.cpp:
```diff
@@ -274,135 +269,130 @@ namespace FEXCore {
+fextl::unique_ptr<ThunkHandler> ThunkHandler::Create() { return fextl::make_unique<ThunkHandler_impl>(); }
 
-    /**
+/**
```
clang-format un-indented only the first line of this docstring. Looks like we'll need to move the other lines manually :(

pmatos replied:
Solved as well.

neobrain commented on FEXCore/Source/Interface/IR/AOTIR.cpp:
```diff
@@ -22,399 +22,383 @@
 namespace FEXCore::IR {
-  AOTIRInlineEntry *AOTIRInlineIndex::GetInlineEntry(uint64_t DataOffset) {
-    uintptr_t This = (uintptr_t)this;
+AOTIRInlineEntry *AOTIRInlineIndex::GetInlineEntry(uint64_t DataOffset) {
```
Seems like the star/ampersand placement is very irregular. It's using `AOTIRInlineEntry *` here (space between type and star) but `auto& [Name, rv] = *reinterpret_cast<ArgsRV_t*>(ArgsRV);`/`HostToGuestTrampolinePtr* TrampolineAddress` in Thunks.cpp. Any idea why it's like that?

I can see why it might choose not to put a space when the type is used in isolation (but then it's not doing that below for the `(AOTIRInlineEntry *)` cast). Also `auto &[Name, rv]` would look kind of weird, though I could get used to it. But everything else should be more consistent.

pmatos replied:
Setting `PointerAlignment` and `ReferenceAlignment` to `Left` aids with consistency, removed `DerivePointerAlignment` that was causing inconsistencies.

lioncash reviewed (commented):
Largely this is very nice! Just a few quick observations.

Also there's a few tables in OpcodeDispatcher that should probably be wrapped in format disabling comments, but I can't actually inline comment on them, because the file is so big that trying to open it locks up the browser tab

neobrain commented on Source/Tools/CommonTools/Linux/Utils/ELFContainer.cpp:
```diff
@@ -69,7 +65,7 @@ ELFContainer::ELFType ELFContainer::GetELFType(int FD) {
 
   // We need to first determine the file size through fstat.
-  struct stat buf{};
+  struct stat buf {};
```
It's adding a space here, but down below it uses `const char *RawString{};` (no space before `{}`).

pmatos replied:
The difference here is that the `{}` after `const char *RawString` are parsed as a Cpp11BracedList and therefore affected by `SpaceBeforeCpp11BracedList`. Setting this option to true, adds a space after `RawString`.

neobrain commented on FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.cpp:
```diff
@@ -6,80 +6,80 @@
-    AddressToEntryMap.clear();
-    UnrelocatedAddressToEntryMap.clear();
+  // Do final code region closures on thread shutdown
+  for (auto &it : AddressToEntryMap) {
```
`auto &it` is used here, but I've seen `auto& var` elsewhere.

pmatos replied:
After latest changes to alignment of references and pointers, everything is to the left.

neobrain reviewed (commented):
After a quick first glance, I like it a lot! There seem to be some inconsistencies in the formatting - maybe a result of clang-unformat overfitting the settings a little? Definitely a big improvement overall though. Other than the obviously inconsistent stuff, I've added few wishlist items, but it's better to wait for others to weigh in before making any changes here.

I do wonder if setting the character limit to a generous 140 is doing more harm than good. You can see a lot of places that were neatly manually split into manual lines before turning into massive blobs. I'm not going to suggest a conservative 80, but maybe 100 is good enough? There's definitely places that are going to suffer from that, so I'm not sure what the right balance is.

neobrain commented on FEXCore/Source/Common/JitSymbols.cpp:
```diff
@@ -7,133 +7,132 @@
+}
+
+void JITSymbols::RegisterNamedRegion(const void *HostAddr, uint32_t CodeSize, std::string_view Name) {
+  if (fd == -1) return;
```
+1 to this, including use of braces everywhere (personal preference), which I believe clang-format can add automatically. Since we're using compact bracing style, it only actually adds one line of code for the closing `}` anyway.

lioncash replied:
I mean like:

```cpp
if (fd == -1) {
  return;
}
```

pmatos replied:
We have quite a few places without braces. Such a change adds braces in many places. I set `InsertBraces: true` and reformated the tree. See https://github.com/FEX-Emu/FEX/pull/3413/commits/652078584b4650958102881904f5f2416929064b

pmatos commented:
Thanks for all the amazing comments until now. I will play a bit with the setup, create a new clang-format and regenerate the tree for further analysis.

pmatos commented on FEXCore/Source/Common/StringConv.h:
```diff
@@ -7,44 +7,42 @@
+  *Result = std::strtoull(Value.data(), nullptr, 0);
+  return true;
+}
+template<typename T, typename = std::enable_if<std::is_enum<T>::value, T>> [[maybe_unused]] static bool Conv(std::string_view Value, T *Result) {
```
This is possible using `AlwaysBreakTemplateDeclarations: true`, and later in clang-format 19 will be renamed. 

I have made the change in commit https://github.com/FEX-Emu/FEX/pull/3413/commits/c2a9b99c48103fa085891fbb45ba865900558d13

pmatos commented on FEXCore/Source/Common/BitSet.h:
```diff
@@ -22,75 +21,49 @@ struct BitSet final {
-  void MemSet(size_t Elements) {
-    memset(Memory, 0xFF, AlignUp(Elements / MinimumSizeBits, MinimumSizeBits));
-  }
+  bool Get(T Element) { return (Memory[Element / MinimumSizeBits] & (1ULL << (Element % MinimumSizeBits))) != 0; }
```
Given the current column limit is set to 140, I believe functions this large will stay single line. Your example is already single line in commit https://github.com/FEX-Emu/FEX/commit/c2a9b99c48103fa085891fbb45ba865900558d13

neobrain replied:
How about `AllowShortFunctionsOnASingleLine: None/Empty` ?

pmatos replied:
We currently have:
`AllowShortFunctionsOnASingleLine: All`
which is what allows functions in a single line. I assume this is what you want? 

@lioncash or were you saying we should never allow functions in the same line? Or maybe just empty ones?

lioncash replied:
I think allowing empty ones would be a decent compromise (I had no idea that was an available sub-option, I thought it was just a straight-up yes/no toggle).

pmatos replied:
OK - So I will allow only empty ones to be collapsed into a single line with `AllowShortFunctionsOnASingleLine: Emtpy`.

pmatos commented:
I have been pushing changes over the old ones as I modify the clang-format file, but formatting changes done over other formatting changes don't reflect the result of doing the changes in one go. I will have to reformat the tree and push-force to see the results of each change.

pmatos commented on FEXCore/Source/Interface/Core/JIT/Arm64/JITClass.h:
```diff
@@ -119,35 +116,32 @@ class Arm64JITCore final : public CPUBackend, public Arm64Emitter  {
-           Shift == IR::ShiftType::LSR ? ARMEmitter::ShiftType::LSR :
-           Shift == IR::ShiftType::ASR ? ARMEmitter::ShiftType::ASR :
-           ARMEmitter::ShiftType::ROR;
+    return Shift == IR::ShiftType::LSL ?
```
Hummm, this is a tricky one. I will leave it until later. :)

pmatos replied:
I think if we want to keep the current style we really need `clang-format off` annotation.

pmatos commented on FEXCore/Source/Interface/Core/ObjectCache/CodeObjectSerializationConfig.h:
```diff
@@ -6,79 +6,81 @@
+
+  bool operator==(CodeObjectSerializationConfig const &other) const {
+    return Cookie == other.Cookie && MaxInstPerBlock == other.MaxInstPerBlock && Arch == other.Arch && MultiBlock == other.MultiBlock &&
+    HardwareTSOEnabled == other.HardwareTSOEnabled && TSOEnabled == other.TSOEnabled && ABILocalFlags == other.ABILocalFlags &&
```
@neobrain is correct. `AlignOperands: Align` is what we want here.

pmatos commented on FEXCore/Source/Interface/Core/OpcodeDispatcher/Crypto.cpp:
```diff
@@ -208,8 +218,11 @@ void OpDispatchBuilder::SHA256MSG1Op(OpcodeArgs) {
 void OpDispatchBuilder::SHA256MSG2Op(OpcodeArgs) {
-  const auto Sigma1 = [this](OrderedNode* W) -> OrderedNode* {
-    return _Xor(OpSize::i32Bit, _Xor(OpSize::i32Bit, _Ror(OpSize::i32Bit, W, _Constant(32, 17)), _Ror(OpSize::i32Bit, W, _Constant(32, 19))), _Lshr(OpSize::i32Bit, W, _Constant(32, 10)));
+  const auto Sigma1 = [this](OrderedNode *W) -> OrderedNode *
+  {
```
It looks as you request in my current version. :)

pmatos commented on FEXCore/Source/Interface/HLE/Thunks/Thunks.cpp:
```diff
@@ -39,31 +39,31 @@ extern "C" {
+  "jmpq *0f(%rip) \n"
+  ".align 8 \n"
+  "0: \n"
+  ".quad 0, 0, 0, 0 \n" // TrampolineInstanceInfo
```
Currently I managed that by tweaking `ContinuationIndentWidth: 2`, so anything that continues a previous expression on a different line is indented 2. Might of course have further consequences.

pmatos commented on FEXCore/Source/Interface/HLE/Thunks/Thunks.cpp:
```diff
@@ -76,134 +76,128 @@ extern char __stop_HostToGuestTrampolineTemplate[];
+  std::shared_mutex ThunksMutex;
+
+  fextl::unordered_map<IR::SHA256Sum, ThunkedFunction*, TruncatingSHA256Hash> Thunks = {
+  {// sha256(fex:loadlib)
```
I think that looks ok atm (in the current changes I will commit soon).

pmatos commented:
With the exception of the clang-format on/off in OpcodeDispatcher and X86Tables tables, I think I have addressed all the comments to the best of my knowledge. Any further comments/opinions/suggestions about the current formatting?

Sonicadvance1 commented:
One thing to note is that we should make sure not to reformat code that isn't ours. `FEXCore/Source/Common/SoftFloat-3e` isn't our code for example.

neobrain commented:
> With the exception of the clang-format on/off in OpcodeDispatcher and X86Tables tables, I think I have addressed all the comments to the best of my knowledge. Any further comments/opinions/suggestions about the current formatting?

Great work! I've been casually browsing over it and nothing else caught my eye so far.

A personal note, I'd like to check how the clang AST matchers used in unittests/ThunkLibs/generator.cpp are formatted now. The readability of matching expressions is very sensitive to reformats, and disabling formatting locally for each of them would add too much noise to each test. We may just need to disable clang-format entirely for this file (though maybe we can do a manual, selective reformat once).

pmatos commented:
> One thing to note is that we should make sure not to reformat code that isn't ours. `FEXCore/Source/Common/SoftFloat-3e` isn't our code for example.

Uff, I expected all code outside external to be ours. I had never seen that directory before. Are there anymore like these outside external I should be aware of?

pmatos commented:
> > One thing to note is that we should make sure not to reformat code that isn't ours. `FEXCore/Source/Common/SoftFloat-3e` isn't our code for example.
> 
> Uff, I expected all code outside external to be ours. I had never seen that directory before. Are there anymore like these outside external I should be aware of?

It looks like Source/Common/cpp-optparse is another example of those.

Sonicadvance1 commented:
> > > One thing to note is that we should make sure not to reformat code that isn't ours. `FEXCore/Source/Common/SoftFloat-3e` isn't our code for example.
> > 
> > 
> > Uff, I expected all code outside external to be ours. I had never seen that directory before. Are there anymore like these outside external I should be aware of?
> 
> It looks like Source/Common/cpp-optparse is another example of those.

I think you got the only two that I remember.

pmatos commented:
> > > One thing to note is that we should make sure not to reformat code that isn't ours. `FEXCore/Source/Common/SoftFloat-3e` isn't our code for example.
> > 
> > 
> > Uff, I expected all code outside external to be ours. I had never seen that directory before. Are there anymore like these outside external I should be aware of?
> 
> It looks like Source/Common/cpp-optparse is another example of those.

A `.clang-format-ignore` file support will be added in clang-18.1.0 (https://github.com/llvm/llvm-project/commit/09308122c6c0fa9eb3d729a2b2909733cbbc2160) but we'll have to maybe work around this until we can use this on our CI.

neobrain commented:
> A personal note, I'd like to check how the clang AST matchers used in unittests/ThunkLibs/generator.cpp are formatted now. The readability of matching expressions is very sensitive to reformats, and disabling formatting locally for each of them would add too much noise to each test. We may just need to disable clang-format entirely for this file (though maybe we can do a manual, selective reformat once).

I took a look at this: Nothing atrocious, however there are some weird effects that don't seem intentional.

Fortunately, the clang AST matcher expressions themselves are formatted fine.

This one is a bit weird though:
```cpp
    CHECK_THAT(
      output,
      matches(classTemplateSpecializationDecl(
        hasName("guest_layout"), hasAnyTemplateArgument(refersToType(asString("struct A"))),
        // The member "data" exists and is defined to a struct...
        has(fieldDecl(
          hasName("data"),
          hasType(hasCanonicalType(hasDeclaration(decl(
            // ... the members of which also use guest_layout (with fixed-size integers)
            has(fieldDecl(hasName("a"), hasType(asString("guest_layout<int32_t>")))), has(fieldDecl(hasName("b"), hasType(asString("guest_"
                                                                                                                                   "layout<"
                                                                                                                                   "int32_"
                                                                                                                                   "t"
                                                                                                                                   ">")))))))))))));
```
Why does it decide to break up the string several times instead of putting any of the `has(...)` expressions on a new line?

Also, [the `runInvocation()` signature](https://github.com/FEX-Emu/FEX/blob/cc9c80d79f6125b0fc2e16739c354cb496498f9b/unittests/ThunkLibs/generator.cpp#L188-L191) gets re-formatted into a single line, overstepping the column limit of 140 characters *a lot*.

Any idea what's going on there?

pmatos commented:
I have just pushed a new change. This one does several things:
* Adds a wrapper to clang-format that takes `.clang-format-ignore` into consideration. It should be easy to use this in our editors instead of the system one. Also CI can use this directly.
* Adds the `.clang-format-ignore` for ease of adding things to ignore.
* Reformatted the tree.

lioncash reviewed (commented):
Doing a quick scan through some files, and this looks quite nice

I think we're missing: `AllowShortFunctionsOnASingleLine: Empty` from previous discussion from being applied, but aside from that, I have no remaining nits =)

pmatos commented:
> Also, [the `runInvocation()` signature](https://github.com/FEX-Emu/FEX/blob/cc9c80d79f6125b0fc2e16739c354cb496498f9b/unittests/ThunkLibs/generator.cpp#L188-L191) gets re-formatted into a single line, overstepping the column limit of 140 characters _a lot_. Any idea what's going on there?

oh my - you're right. It looks like a bug in clang-format. The new line has something like 195 characters. I will look into.

pmatos commented:
> Doing a quick scan through some files, and this looks quite nice
> 
> I think we're missing: `AllowShortFunctionsOnASingleLine: Empty` from previous discussion from being applied, but aside from that, I have no remaining nits =)

I think you're right, sorry. I will look into the issue @neobrain pointed out and then re-run a format taking that into consideration.

neobrain commented on unittests/FEXLinuxTests/tests/signal/invalid_hlt.cpp:
```diff
@@ -12,14 +12,13 @@
 #pragma GCC diagnostic ignored "-Wattributes" // Suppress warning in case control-flow checks aren't enabled
-__attribute__((naked, nocf_check))
-  static void InvalidINT() {
+__attribute__((naked, nocf_check)) static void InvalidINT() {
```
Should we enable `BreakAfterAttributes`?

CC @lioncash , since this seems analogous to breaking after `template<>`.

If we do this, we may want to add these macros to `AttributeMacros`: `JEMALLOC_NOTHROW`, `FEX_ALIGNED`, `FEX_ANNOTATE`, `FEX_DEFAULT_VISIBILITY`, `FEX_NAKED`, `FEX_PACKED`, `FEXCORE_PRESERVE_ALL_ATTR`, `GLIBC_ALIAS_FUNCTION`

neobrain commented on unittests/ThunkLibs/generator.cpp:
```diff
@@ -376,564 +372,506 @@ Fixture::GenOutput Fixture::run_thunkgen(std::string_view prelude, std::string_v
+          hasName("data"),
+          hasType(hasCanonicalType(hasDeclaration(decl(
+            // ... the members of which also use guest_layout (with fixed-size integers)
+            has(fieldDecl(hasName("a"), hasType(asString("guest_layout<int32_t>")))), has(fieldDecl(hasName("b"), hasType(asString("guest_"
+                                                                                                                                   "layout<"
+                                                                                                                                   "int32_"
+                                                                                                                                   "t"
+                                                                                                                                   ">")))))))))))));
```
Following up on https://github.com/FEX-Emu/FEX/pull/3413#issuecomment-1959081683:
> Why does it decide to break up the string several times instead of putting any of the has(...) expressions on a new line?

`PenaltyBreakString: 10` will fix this, but I haven't checked what ripple effects that has on the rest of FEX. Seems highly desirable in this particular instance though. (Note that clang-format won't un-break strings, so changes here need to be re-applied on the pre-format sources).

pmatos replied:
I have removed the option and compared the trees with and without the option side by side and it seems that there's no situation where that option actually makes things better.

The only different I see that's more relevant are things like (in `eflags_signal.cpp`) for example:

![Screenshot from 2024-02-23 09-26-20](https://github.com/FEX-Emu/FEX/assets/7911/18abaa1f-4790-400c-b46b-b23f25612064)

I will regenerate the tree again.

pmatos commented:
There has not been any comments as of late on this. Here's my suggestion:
1. I will work on CI integration in a different PR.
2. Choose a day for the merge of the reformatting.
3. On that day, we commit:
  * the reformatting, then 
  * .clang-format file and 
  * the ignore blame list file (which contains the sha of the reformatting commit)
  * and finally the CI integration PR.

What do you think?

Sonicadvance1 commented:
> There has not been any comments as of late on this. Here's my suggestion:
> 
> 1. I will work on CI integration in a different PR.
> 2. Choose a day for the merge of the reformatting.
> 3. On that day, we commit:
> 
> * the reformatting, then
> * .clang-format file and
> * the ignore blame list file (which contains the sha of the reformatting commit)
> * and finally the CI integration PR.
> 
> What do you think?

Sounds reasonable to me.

neobrain commented:
> What do you think?

Yes, really happy with how the formatting turned out - let's move forward with it :)

alyssarosenzweig commented:
We probably want to prioritize reviewing & landing as much as possible this week then, else the conflicts are going to stink. (There are ways around this but it's annoying.)

pmatos commented:
> We probably want to prioritize reviewing & landing as much as possible this week then, else the conflicts are going to stink. (There are ways around this but it's annoying.)

Yes - that makes sense.  We can always try to do that this week. If it doesn't work, we can push this plan to next week instead. I think there's not a lot of pressure on this. :)

pmatos commented:
As mentioned in discord:

I added in #3572, a script to help reformat given a clang-format-ignore (cause there are files we don't want to reformat).
For people working on branches, before rebasing on main, it's easier to reformat the whole tree first locally.
This should reduce conflicts.
Once the massive refactor is landed I can add another PR to ensure that this commit is ignore in git blames but I need it's sha first.
So there'll be 4 commits in total, in landing order:

* Reformat #3571
* git blame ignore  - to come
* clang-format file #3572
* CI workflow #3561

Once #3571 lands, get the files from #3572 and run on your local branches:
`find . -iname '*.h' -o -iname '*.cpp' -exec python3 Scripts/clang-format.py -i \{\} \;`

This should help with conflicts. Make sure you have a recent clang-format.

pmatos commented:
Reformat has landed - so gitblame ignore revs file is in #3575 .

pmatos commented:
The last PR from this work was just landed. Thank you all for the discussion, reviewing and landing the PRs. Closing this.
